siliconcompiler/core.py

```diff
@@ -3419,8 +3429,12 @@ def edge(self, flow, tail, head, tail_index=0, head_index=0):
                 self.error(f'Illegal step name: {step} is reserved')
                 return
 
-        # Adding
-        self.add('flowgraph', flow, head, str(head_index), 'input', (tail, str(tail_index)))
+        tail_node = (tail, str(tail_index))
+        if tail_node in self.get('flowgraph', flow, head, str(head_index), 'input'):
+            self.logger.warning(f'Edge from {tail}{tail_index} to {head}{head_index} already exists, skipping')
+            return
+
+        self.add('flowgraph', flow, head, str(head_index), 'input', tail_node)
```

Comment:
why not cast the head and tail index to str() at the start of this method so the types are consistent throughout?

Reply:
Just a copy-paste thing, but I'll clean it up - nice suggestion.